lib/src/pubspec.dart

```diff
@@ -102,19 +100,75 @@ class Pubspec {
 
   String get issueTracker => _inner.issueTracker?.toString();
 
-  bool get hasDartSdkConstraint => _inner.environment.containsKey('sdk');
+  bool get hasDartSdkConstraint => sdkConstraintStatus.hasConstraint;
 
-  bool get shouldWarnDart2Constraint {
-    final vc = _inner.environment['sdk'];
-    if (vc == null) {
-      return false;
-    }
-    return !vc.allows(_dart2Last);
-  }
+  bool get shouldWarnDart2Constraint => !sdkConstraintStatus.enablesDart2Latest;
 
   bool get hasGitDependency =>
       _inner.dependencies.values.any((d) => d is GitDependency);
 
   bool get hasUnrestrictedGitDependency => _inner.dependencies.values
       .any((d) => d is GitDependency && (d.ref == null || d.ref.length < 40));
+
+  SdkConstraintStatus get sdkConstraintStatus =>
+      SdkConstraintStatus.fromSdkVersion(_inner.environment['sdk']);
+}
+
+final _range0 = VersionConstraint.parse('<1.0.0');
+final _range1 = VersionConstraint.parse('>=1.0.0 <2.0.0');
+final _range2 = VersionConstraint.parse('>=2.0.0 <3.0.0');
+final _range2Latest = VersionConstraint.parse('>=2.9999.0 <3.0.0');
+final _futureRange = VersionConstraint.parse('>=3.0.0');
+
+/// Detailed support coverage for the SDK constraint.
+class SdkConstraintStatus {
+  /// Whether it is non-empty, bounded constraint.
+  final bool hasConstraint;
+
+  /// Whether it allows anything from the <1.0.0 range.
+  final bool enablesDart0;
```

Comment:
Do we need this information anywhere? Otherwise I suggest we shouldn't expose it...

Reply:
It is useful for completeness, but we could remove it, because I cannot think of a constraint that enables <1.0 and 2.0 but not 1.0, so `enablesDart1` should cover for `isDart2Compatible`.